Provide API to filter unwanted contributions

tortmayr opened the pull request:
#### What it does

Adds the `ContributionFilterRegistry` which can be used to filter contributions of Theia extensions before they are bound.

This mechanism can be used by application developers to specifically filter individual contributions from Theia extensions they have no control over, i.e. core or 3rd party extensions.

Resolves #9069

Contributed on behalf of STMicroelectronics
Signed-off-by: Tobias Ortmayr <tortmayr@eclipsesource.com>

#### How to test

- You can take a look at the included test case, or
- Contribute a `ContributionFilter` in an example Theia application

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 
+## v1.13.0
+
+- [core] add API to filter contributions at runtime [#9317](https://github.com/eclipse-theia/theia/pull/9317) Contributed on behalf of STMicroelectronics
```
Is there a strong requirement to include `contributed on behalf...` in the changelog, it does not follow the format of existing entries. In the past it was simply referenced in the pull-request description itself, and/or in the commit.

JonasHelming replied:
Yes it is

marcdumais-work replied:
This was previously agreed-upon, as one of the ways to compensate, for the lost of visibility for sponsoring contribution, compared to the "old" way of adding a mention of said contributions, as part of the copyright header of each file. 

Another permissible way is to add a similar mention in the body of commit messages.

vince-fugnitto commented on examples/api-samples/src/browser/contribution-filter/sample-filtered-command-contribution.ts:
```diff
@@ -0,0 +1,50 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
```
minor: please include a newline after copyright headers.

vince-fugnitto commented on examples/api-samples/src/browser/contribution-filter/sample-filtered-command-contribution.ts:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { Command, CommandContribution, CommandRegistry, ContributionFilter, equals, NameBasedContributionFilter } from '@theia/core/lib/common';
+import { injectable, interfaces } from '@theia/core/shared/inversify';
+export namespace SampleFilteredCommand {
```
minor: add newline.

vince-fugnitto commented on packages/core/src/common/contribution-filter/contribution-filter-registry.ts:
```diff
@@ -0,0 +1,69 @@
+import { ContributionFilter, ContributionType } from './contribution-filter';
+import { applyFilters } from './filter';
+
+const GENERIC_CONTRIBUTION_FILTER_KEY = '*';
```
I believe this can be moved in the class itself if necessary.

marechal-p commented on packages/core/src/common/contribution-filter/string-based-filter.ts:
```diff
@@ -0,0 +1,106 @@
+        testStr = testStr.toLowerCase();
+        patterns = patterns.map(pattern => pattern.toLowerCase());
+    }
+    return patterns.find(pattern => testStr.includes(pattern)) !== undefined;
```
`.some` instead of `.find`, this way you don't need to do `!== undefined`.

marechal-p reviewed (commented):
Thanks for the contribution! Since this is going to be something core to Theia, I wanted to properly review it. Unfortunately I found enough things that bothered that I created a branch with most changes I'd be more comfortable with, see: https://github.com/eclipsesource/theia/compare/filter-api...eclipse-theia:mp/filter-api-2

My issues mostly revolve around:

1. Class hierarchies too deep for not much value. Take the following:

    ```ts
    @injectable()
    export class SampleFilteredCommandContributionFilter extends NameBasedContributionFilter {
        contributions = [CommandContribution];
        doTest(toTest: string): boolean {
            return equals(toTest, false, 'SampleFilteredCommandContribution');
        }
    }
    ```

    It takes 3 abstract classes just not to write:

    ```ts
            return equals(toTest.constructor.name, false, 'SampleFilteredCommandContribution');
    ```

    It also seems to me that filters won't really be extended in such a way that we'll need crazy extensibility. Correct me if I am wrong.

2. Some of the logic can be rewritten more concisely.
3. Some styling issues. It most likely means that our linting rules are not good enough?

If you agree with the changes I proposed on the `mp/filter-api-2` branch, feel free to squash them into your commit, populating the `Co-Authored-By: Paul ...` line for IP stuff.

tsmaeder commented on packages/core/src/common/contribution-filter/contribution-filter-registry.ts:
```diff
@@ -0,0 +1,69 @@
+        });
+    }
+
+    private addFilter(type: ContributionType, filter: ContributionFilter): void {
```
> 
> minor: please place `protected ` and `private` methods at the end of the class (after public methods).

Not sure that is a good policy in general: often, private methods are used to structure larger, public methods: in that case, I prefer to have them close to where they are used.

tsmaeder commented on packages/core/src/common/contribution-filter/contribution-filter.ts:
```diff
@@ -0,0 +1,40 @@
+     * Contribution types for which this filter is applicable. If `undefined` or empty this filter
+     * will be applied to all contribution types.
+     */
+    contributions?: ContributionType[];
```
The contribution type should not be part of the `ContributionFilter` interface: it makes it impossible to reuse the same filter implementation for multiple contributions. We can just remove this interface and rewrite `FilterRegistry.addFilter` like so:
```
addFilter(filter: Filter, restrictTo?: ContributionType[]);
```` 
if you don't pass `restrictTo`, it applies to all contributions. More philosophically, it does not feel to me that the contribution type is part of the state of the Filter, it's part of the state of the registry. On indication of that is that once the filter is added to the registry, the "contributions" field becomes unused.
I think the base problem is that we're missing a "ContributionFilterContribution" type.

marechal-p replied:
You can very well reuse filters right now: the `ContributionFilter` type is just a `Filter<Object>` with that `contributions` field. So if you want to reuse a filter, you would implement `Filter<Object>` and subclass it as many times as you want in various `ContributionFilter` implementations. I think this pattern is "forced" upon us to work with Inversify... It's not so much that `contributions` is really part of the state of the contribution filter, but rather it is used to communicate with the registry through Inversify injection.

Ultimately the issue is that `ContributionFilter` are injectable classes. This makes it hard to pass around a `Filter` constructor to resolve without the list of contributions it applies to.

The `ContributionFilterContribution` could be a different API indeed.

marechal-p replied:
@tsmaeder here's an attempt at making it more functional: https://github.com/eclipsesource/theia/compare/filter-api...eclipse-theia:mp/filter-api-4

See [sample-filtered-command-contribution.ts](https://github.com/eclipse-theia/theia/blob/7d860c78446780de7e04465ef9c387c24db7ddab/examples/api-samples/src/browser/contribution-filter/sample-filtered-command-contribution.ts#L48-L65) for usage example. I had to use a different kind of contribution mechanism, one closer to what we currently do for commands.

tortmayr replied:
I'm in favor of a more functional approach and would like to squash the proposal from @marechal-p.
@tsmaeder Do you have any objections or additional input?

tortmayr commented:
Thanks @marechal-p for your in-depth review and the proposed changes.I squashed your changes and added the `Co-Authored-By:` line to the commit message. I believe all mentioned issues should be resolved now.

tsmaeder reviewed (commented):
This is a valuable contribution, but I think it would benefit from being done in a more functional style.

marechal-p commented on packages/core/src/common/contribution-filter/filter.ts:
```diff
@@ -0,0 +1,48 @@
+ * from a set of objects. The `test` function can be applied to an object
+ * of matching type and returns `true` if the object should be filtered out.
+ */
+export interface Filter<T extends Object> {
```
`Object` is a catch all. It is different from `object`. Almost everything except `null` and `undefined` are `Object`. It allows typing to know about `Object.constructor`.

marechal-p replied:
I think `Filters` being an interface can be more convenient since it makes holding state easier if needed. At least it sounds more convenient than closures. Either would work.

tsmaeder replied:
:shrug: Wat?! ;-)

tsmaeder replied:
> 
> 
> I think `Filters` being an interface can be more convenient since it makes holding state easier if needed. At least it sounds more convenient than closures. Either would work.

I disagree here philosophically: an interface should pose minimal restrictions on it's implementers. If all we need to do is call a function, that's what we should require.

tortmayr commented:
@paul-marechal  I have squashed your proposal for a more functional filtering approach and rebased the PR onto the latest master.

marcdumais-work commented:
@tortmayr  just so you know, Paul is OOO this week, so it could take a few business days for a response, from him.

paul-marechal commented on packages/core/src/common/contribution-provider.ts:
```diff
@@ -73,3 +82,17 @@ export function bindContributionProvider(bindable: Bindable, id: symbol): void {
         .toDynamicValue(ctx => new ContainerBasedContributionProvider(id, ctx.container))
         .inSingletonScope().whenTargetNamed(id);
 }
+
+/**
+ * Helper function to bind a service to a list of contributions easily.
+ * @param bindable a Container or the bind function directly.
+ * @param service an already bound service to refer the contributions to.
+ * @param contributions array of contribution identifiers to bind the service to.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function bindContribution(bindable: Bindable, service: interfaces.ServiceIdentifier<any>, contributions: interfaces.ServiceIdentifier<any>[]): void {
+    const bind: interfaces.Bind = Bindable.isContainer(bindable) ? bindable.bind.bind(bindable) : bindable;
+    for (const contribution of contributions) {
+        bind(contribution).toService(service);
+    }
+}
```
Scratch that, found a use: https://github.com/eclipse-theia/theia/pull/9317/files#diff-eb37d33bb557047ff9e07f8f6dd0281f0b7bde0c2cad08275876a334710928d4R70

tortmayr commented on packages/core/src/common/contribution-filter/string-utils.ts:
```diff
@@ -0,0 +1,60 @@
+ * @param ignoreCase Flag, to indicate wether the test should be case-sensitive or not
+ * @param patterns  The set of patterns that should be tested for inclusion
+ * @returns `true` if any of the given patterns is included in the test string, `false` otherwise
+ */
+export function includes(testStr: string, ignoreCase: boolean, ...patterns: string[]): boolean {
+    if (ignoreCase) {
+        testStr = testStr.toLowerCase();
+        patterns = patterns.map(pattern => pattern.toLowerCase());
+    }
+    return patterns.some(pattern => testStr.includes(pattern));
+}
+
+/**
+ * Utility function to test wether any of the given string patterns is equal to the
+ * tested string.
+ * @param testStr String that should be tested
+ * @param ignoreCase Flag, to indicate wether the test should be case-sensitive or not
+ * @param patterns  The set of patterns that should be tested for equality
+ * @returns `true` if any of the given patterns is equal to the test string, `false` otherwise
+ */
+export function equals(testStr: string, ignoreCase: boolean, ...patterns: string[]): boolean {
+    if (ignoreCase) {
+        testStr = testStr.toLowerCase();
+        patterns = patterns.map(pattern => pattern.toLowerCase());
+    }
+    return patterns.some(pattern => testStr === pattern);
+
+}
+
+/**
+ * Utility function to test wether a string matches any of the given regular expression patterns.
+ * @param testStr String that should be tested
+ * @param ignoreCase Flag, to indicate wether the test should be case-sensitive or not
+ * @param patterns The set of regular expressions that should be matched
+ * @returns `true` if the test string matches any of the given regular expressions, `false` otherwise
+ */
+export function matches(testStr: string, ignoreCase: boolean, ...patterns: (RegExp | string)[]): boolean {
+    const flags = ignoreCase ? 'i' : undefined;
+    return patterns.some(pattern => new RegExp(pattern, flags).test(testStr));
+}
```
:+1: I removed this file

tortmayr commented:
Thanks for the re-review @paul-marechal. I removed the `string-utils.ts` file and rebased onto the current master

tsmaeder commented on packages/core/src/common/contribution-provider.ts:
```diff
@@ -48,10 +50,17 @@ class ContainerBasedContributionProvider<T extends object> implements Contributi
                 currentContainer = recursive === true ? currentContainer.parent : null;
             }
-            this.services = currentServices;
+            if (filterRegistry === undefined) {
```
Why not just do 
```
if (filterRegistry) {
} else {
}
```
We know it's an object, so if it's truthy, we can use it.

tsmaeder commented on packages/core/src/common/contribution-filter/filter.ts:
```diff
@@ -0,0 +1,43 @@
+ * @param negate Negation flag. If set to true the result of all filters is negated
+ * @returns The set of filtered arguments
+ */
+export function applyFilters<T extends Object>(toFilter: T[], filters: Filter<T>[], negate: boolean = false): T[] {
```
This function seems to be used exclusively in `ContributionFilterRegistryImpl`: I would move it to that class. Let's not create public API unless needed.

tsmaeder replied:
We could also lose the `negate` then, which seems confusing.

tsmaeder commented on packages/core/src/common/contribution-filter/contribution-filter-registry.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    addFilters(types: '*' | ContributionType[], filters: Filter<Object>[]): void {
+        if (this.initialized) {
```
I'm not sure we do this check for any other implementations of the registry/contribution callback pattern. Also: I'm not seeing an `this.initialized = true` here ;-)

paul-marechal replied:
```ts
    constructor(
        @multiInject(FilterContribution) @optional() contributions: FilterContribution[] = []
    ) {
        for (const contribution of contributions) {
            contribution.registerContributionFilters(this);
        }
        this.initialized = true;
    }
```

tsmaeder replied:
I understand that, but we don't seem to take this precaution in the other instances of this pattern, so why start now and why stop here?

paul-marechal replied:
Some registries might be designed to be mutated at any time, this one cannot (more precisely it wouldn't have an effect and lead to ignored filters which could be confusing). I'm all for failing as early as possible, so if other places benefit from such checks then yea we should do this there too. But this work is out of scope of this PR I think.

paul-marechal replied:
At the end of the day it depends on what's the purpose of the component right, in this case it would cause issues to mutate later hence why I locked it with the `initialized` flag.

tsmaeder replied:
Fair enough.

tsmaeder commented on packages/core/src/common/contribution-filter/contribution-filter-registry.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    protected getOrCreate(type: ContributionType): Filter<Object>[] {
+        let value = this.typeToFilters.get(type);
```
Why not just redo this as an "addFilters" method? Do you see potential for extension of this class?

paul-marechal replied:
I wrote that to simplify the writing/reading of `addFilter`.

tsmaeder reviewed (changes requested):
I like the general shape of this PR a lot now. I have a couple of questions I feel need answers though, hence a "Request changes".

tsmaeder commented:
@tortmayr @paul-marechal so how are we going to proceed here?

paul-marechal commented:
Waiting for @tortmayr to address the unaddressed comments?

tortmayr commented:
@paul-marechal I have addressed the open comments and rebased to the latest master.

JonasHelming commented:
@tsmaeder @paul-marechal : Are you good with this PR now?

tsmaeder reviewed (approved):
Looks good to me now.

paul-marechal reviewed (approved):
LGTM too, thanks!
